### PR TITLE
Add first version of C4 architecture diagrams

### PR DESCRIPTION
**Pull request by ChristopherSpelt**
## Description

Creates a `docs/architecture/ folder containing: 

- `architecture.md`: a Markdown document containing C4-style diagrams for the Algorithm Management Toolkit. This document contains a system diagram, a container diagram and a component diagram, and a walkthrough and example text explaining the diagrams. 

Resolves: https://github.com/orgs/MinBZK/projects/7/views/3?pane=issue&itemId=59900118

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [x]  I have followed the project's coding conventions and style guidelines.
- [x]  I have rebased my branch onto the latest commit of the main branch.
- [x]  I have squashed or reorganized my commits into logical units.
- [x]  I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

### CODE REVIEW

**Comment by GuusjeJuijn** [comment]
##Description

Creates a `docs/architecture/ folder containing: 

- `architecture.md`: a Markdown document containing C4-style diagrams for the Algorithm Management Toolkit. This document contains a system diagram, a container diagram and a component diagram, and a walkthrough and example text explaining the diagrams. 

Resolves: https://github.com/orgs/MinBZK/projects/7/views/3?pane=issue&itemId=59900118

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [ ]  I have tested the changes locally and verified that they work as expected.
- [ ]  I have followed the project's coding conventions and style guidelines.
- [ ]  I have rebased my branch onto the latest commit of the main branch.
- [ ]  I have squashed or reorganized my commits into logical units.
- [ ]  I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

**Comment by laurensWe** [comment]
First very minor remark, could you `rebase -i` and then `fixup` the last two commits?

File: docs/architecture/architecture.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# Algorithm Management Toolkit Architecture
+
+This document contains architectural decisions related to the Algorithm Managment Toolkit.
+
+## System Context
+
+The Algorithm Management Toolkit is an application that users can use to complete requirements
+that are specified by the algoritmekader. The diagram below sketches the broader system context
```
Comment (laurensWe) [inline docs/architecture/architecture.md:8]: I would add here also that the extra context on the other side of the the AMT is of the Algorithm Register because in the end the goal of this tool is to output a system card, this whole part i miss a bit

File: docs/architecture/architecture.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+UpdateRelStyle(Tasks, InstrumentRegister, $offsetY="10", $offsetX="-150")
+
+```
+
+### Walkthrough
+Suppose a user wants to perform a specific task from the Algoritmekader. To execute this task, the user has 2 options.
+
+The first option is to execute the task locally by using the command line interface tool (CLI). The CLI tool imports the task from a Python library named `Tasks`. This library contains executable tasks that implement the measures and instruments from the `Instrument Register` (which are specified in the Algoritmekader). Instructions on how to perform these tasks are imported from the Instrument Register. There exists a one-to-one correspondence between measures and instruments in the `Instrument Register` and the task's within the `Tasks` library.
+
+The second option is to use the Algorithm Management Toolkit (AMT). The user starts by visiting the Algorithm Management Toolkit website. Here, the user encounters a front end interface showing a planning board for projects and tasks. This planning board contains 3 columns: ‘To do’,  ‘Doing’ and ‘Done’. When a user drags a task from ’To do’ to ‘Doing’, the front end makes an API call to the back end of the AMT.
```
Comment (laurensWe) [inline docs/architecture/architecture.md:134]: We haven't really decided on the columns of the kanban board though, it is a great start but also we haven't included the lifecyle of the system/model yet but also maybe we would need a column Blocked or Review. I just want to write it somewhere down here.

File: docs/architecture/architecture.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+The back end consists of three components, showed in the component diagram at the end of this page.
```
Comment (laurensWe) [inline docs/architecture/architecture.md:136]:
```suggestion
The backend consists of three components, showed in the component diagram at the end of this page.
```

File: docs/architecture/architecture.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+1. An API application, which provides the project and tasks management functionality via HTTPS
```
Comment (laurensWe) [inline docs/architecture/architecture.md:137]:
```suggestion
1. An API application, which provides the project and tasks management functionality via HTTPS.
```

File: docs/architecture/architecture.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+2. The business logic, which is the core logic of the AMT
```
Comment (laurensWe) [inline docs/architecture/architecture.md:138]:
```suggestion
2. The business logic, which is the core logic of the AMT.
```

File: docs/architecture/architecture.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+3. A system state, which provides the state of the AMT.
+
+When receiving an API call, the API application forwards the instruction to the business logic. The business logic, in turn, updates the system state and submits the task to the Redis message queue. The message queue stores the task messages until a Celery worker is ready to process a specific task. When a Celery worker is available, it uses the task library to execute the task. After the task is completed by the worker, the result is sent back to the business logic via the message queue. The business logic now sends an update to the system state and writes the result to the database. Finally, the business logic writes a System Card to an external repository.
```
Comment (laurensWe) [inline docs/architecture/architecture.md:141]:
```suggestion
When receiving an API call, the application forwards the instruction to the business logic. The business logic, in turn, updates the system state and submits the task to the Redis message queue. The message queue stores the task messages until a Celery worker is ready to process a specific task. When a Celery worker is available, it uses the task library to execute the task. After the task is completed by the worker, the result is sent back to the business logic via the message queue. The business logic now sends an update to the system state and writes the result to the database. Finally, the business logic writes a System Card to an external repository.
```

File: docs/architecture/architecture.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        System(Repository, "Repository", "Contains a System Card with filled in measures and instruments")
+    }
+
+    Person(user0, "User", "User wants to comply with regulations and required
+    ethical frameworks")
+
+
+    Rel(Algoritmekader, TAD, "Passes required measures and instruments")
+    UpdateRelStyle(Algoritmekader, TAD, $offsetY="50", $offsetX="-150")
+    UpdateLayoutConfig($c4ShapeInRow="3", $c4BoundaryInRow="1")
```
Comment (laurensWe) [inline docs/architecture/architecture.md:40]: i would maybe have shapeinrow=2 instead of 3, as the text is falling behind the boundary and also that we have everything except the AMT is not blue.

File: docs/architecture/architecture.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+UpdateRelStyle(user0, FrontEnd, $offsetY="-40", $offsetX="-70")
+
+Rel(CLI, Tasks, "Imports tasks from")
+UpdateRelStyle(CLI, Tasks, $offsetY="-5", $offsetX="-120")
+
+BiRel(FrontEnd, Backend, "Front End makes API calls, <br/> Back End sends live updates", "HTTPS, websocket")
+UpdateRelStyle(FrontEnd, Backend, $offsetY="-10", $offsetX="-165")
+
+
+BiRel(Backend, Db, "Reads form and <br/>writes to", "")
```
Comment (uittenbroekrobbert) [inline docs/architecture/architecture.md:107]: Typo fix 

```suggestion
BiRel(Backend, Db, "Reads from and <br/>writes to", "")
```

File: docs/architecture/architecture.md (hunk 2 above)
Comment (uittenbroekrobbert) [reply docs/architecture/architecture.md:134]: We could reword it to be more general.

The planning boards contains columns which corresponds to the status of a task, like 'ToDo', 'Doing' and 'Done'. When a user drags a task to re-order tasks or another column, the front-end makes an API call to the back-end of the AMT to update the task accordingly.

File: docs/architecture/architecture.md (hunk 3 above)
Comment (uittenbroekrobbert) [reply docs/architecture/architecture.md:136]: I noticed that throughout the document "back end" is used as 2 words.

**Review by uittenbroekrobbert — COMMENTED** [review]
LGTM! Made some small remarks.

File: docs/architecture/architecture.md (hunk 1 above)
Comment (GuusjeJuijn) [reply docs/architecture/architecture.md:8]: Good point, I will add that to the text + to the system context diagram!

File: docs/architecture/architecture.md (hunk 7 above)
Comment (GuusjeJuijn) [reply docs/architecture/architecture.md:40]: good point :)

File: docs/architecture/architecture.md (hunk 2 above)
Comment (GuusjeJuijn) [reply docs/architecture/architecture.md:134]: done

**Review by laurensWe — APPROVED** [review]
LGTM :)
